File: overlay_posture_from_csv/overlay_posture_from_csv.m
% Set the default renderer
set(0, 'DefaultFigureRenderer', 'painters');

% Define the color palette
palette.color.ir = [249,82,91]/255;
palette.color.ird = [134,46,56]/255;
palette.color.bd = [25,211,197]/255;
palette.color.bl = [138,210,211]/255;
palette.color.yd = [252,176,33]/255;
palette.color.yl = [254,197,87]/255;
palette.color.vy = [240,223,0]/255;
palette.color.db = [49,51,53]/255;
palette.color.cg11 = [84,86,91]/255;
palette.color.cg9 = [117,118,121]/255;
palette.color.cg7 = [150,152,153]/255;
palette.color.cg4 = [187,186,186]/255;
palette.color.cg1 = [219,219,221]/255;

% Set the default color palette
palette.default_color = [palette.color.ir;
    palette.color.bd;
    palette.color.yd;
    palette.color.db;
    palette.color.bl;
    palette.color.yl];

% Select the CSV file using a file dialog
[csvFilename, csvPathname] = uigetfile("*.csv", "Select a CSV file to parse");

% Check if the user selected a file or canceled the dialog
if isequal(csvFilename, 0) || isequal(csvPathname, 0)
    disp('User canceled the CSV file selection');
    return;
end

% Read the entire CSV file without skipping any rows
fileID = fopen(fullfile(csvPathname, csvFilename));
rawData = textscan(fileID, '%s', 'Delimiter', '\n');
fclose(fileID);
rawData = rawData{1};

% Extract the body part names from the second row
bodyParts = strsplit(rawData{2}, ',');
bodyParts = matlab.lang.makeValidName(bodyParts);

% Read the data starting from the fourth line
dataStartLine = 4;
dataLines = rawData(dataStartLine:end);

% Prompt the user to select a body part from a dropdown menu
[selectionIndex, ok] = listdlg('PromptString', 'Select a body part:', ...
                               'SelectionMode', 'single', ...
                               'ListString', bodyParts(2:3:end));

if ~ok
    disp('User canceled the body part selection');
    return;
end

selectedBodyPart = bodyParts{(selectionIndex - 1) * 3 + 2};

% Initialize arrays for coordinates and likelihood
numLines = length(dataLines);
xCoords = zeros(numLines, 1);
yCoords = zeros(numLines, 1);
likelihood = zeros(numLines, 1);

% Loop through each data line to extract x, y coordinates and likelihood
for i = 1:numLines
    lineData = strsplit(dataLines{i}, ',');
    xCoords(i) = str2double(lineData{(selectionIndex - 1) * 3 + 2});
    yCoords(i) = str2double(lineData{(selectionIndex - 1) * 3 + 3});
    likelihood(i) = str2double(lineData{(selectionIndex - 1) * 3 + 4});
end

% Normalize likelihood to [0, 1] for color mapping
normalizedLikelihood = (likelihood - min(likelihood)) / (max(likelihood) - min(likelihood));

% Generate colors from red to grey
startColor = palette.color.ir;
endColor = palette.color.cg1;
colors = (1 - normalizedLikelihood) * startColor + normalizedLikelihood * endColor;

% Select the video file using a file dialog
[videoFilename, videoPathname] = uigetfile("*.avi;*.mp4", "Select a video file");

% Check if the user selected a file or canceled the dialog
if isequal(videoFilename, 0) || isequal(videoPathname, 0)
    disp('User canceled the video file selection');
    return;
end

% Read the video
videoReader = VideoReader(fullfile(videoPathname, videoFilename));

% Prompt the user to select the destination folder
destinationPath = uigetdir('', 'Select the destination folder');

% Check if the user selected a folder or canceled the dialog
if isequal(destinationPath, 0)
    disp('User canceled the folder selection');
    return;
end

% Generate the output video filename
[~, name, ~] = fileparts(videoFilename);
outputVideoFilename = fullfile(destinationPath, [name, '_overlayed_', selectedBodyPart, '.avi']);
videoWriter = VideoWriter(outputVideoFilename);
open(videoWriter);

% Create a figure for plotting
hFig = figure('Visible', 'off');

% Loop through each frame of the video
for k = 1:videoReader.NumFrames
    frame = read(videoReader, k);
    imshow(frame, 'Border', 'tight');
    hold on;
    
    % Plot the body part with the color corresponding to its likelihood
    if k <= numLines
        plot(xCoords(k), yCoords(k), 'o', 'MarkerEdgeColor', 'k', 'MarkerFaceColor', colors(k, :), 'MarkerSize', 12, 'LineWidth', 2);
    end
    
    % Get the frame with the overlay
    overlayFrame = getframe(hFig);
    writeVideo(videoWriter, overlayFrame);
    hold off;
end

% Close the video writer
close(videoWriter);
close(hFig);

disp(['Overlay video saved to: ', outputVideoFilename]);

File: overlay_posture_from_csv/overlay_posture_from_csv.py
import cv2
import pandas as pd
import numpy as np
from tqdm import tqdm
import argparse

def read_csv(filename):
    df = pd.read_csv(filename, header=[1, 2])
    body_parts = df.columns.levels[0].tolist()[1:]  # Ignore scorer column
    return df, body_parts

def normalize_likelihood(df):
    min_val = df.min().min()
    max_val = df.max().max()
    return (df - min_val) / (max_val - min_val)

def generate_color_palette(df):
    start_color = np.array([249, 82, 91, 255])  # Red
    end_color = np.array([219, 219, 221, 255])  # Grey
    colors = np.empty((len(df), 4), dtype=np.uint8)
    
    for i in range(len(df)):
        val = df.iloc[i].values[0]
        colors[i] = (1.0 - val) * start_color + val * end_color
    return colors

def overlay_bodypart_on_video(video_file, output_file, df, colors):
    cap = cv2.VideoCapture(video_file)
    fourcc = cv2.VideoWriter_fourcc(*'XVID')
    out = cv2.VideoWriter(output_file, fourcc, 20.0, (int(cap.get(3)), int(cap.get(4))))

    for frame_idx in tqdm(range(len(df))):
        ret, frame = cap.read()
        if not ret:
            break

        for part, color in zip(df.columns.levels[0][1:], colors):
            x, y = int(df[(part, 'x')].iloc[frame_idx]), int(df[(part, 'y')].iloc[frame_idx])
            if 0 <= x < frame.shape[1] and 0 <= y < frame.shape[0]:
                cv2.circle(frame, (x, y), 6, color.tolist(), 2)
        
        out.write(frame)

    cap.release()
    out.release()

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Overlay body part coordinates on video.')
    parser.add_argument('csv_file', type=str, help='Path to the CSV file')
    parser.add_argument('video_file', type=str, help='Path to the video file')
    parser.add_argument('output_file', type=str, help='Path to the output video file')
    args = parser.parse_args()

    df, body_parts = read_csv(args.csv_file)
    likelihood_df = df.xs('likelihood', level=1, axis=1)
    normalized_likelihood_df = normalize_likelihood(likelihood_df)
    colors = generate_color_palette(normalized_likelihood_df)

    overlay_bodypart_on_video(args.video_file, args.output_file, df, colors)